src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,211 @@
+le_antisymm (card_subtype_mono _ _ (λ _ h H, h (hk _ H))) (ham_dist_smul_le _ _ _)
+
+lemma ham_dist_eq (x y : ham β) : ham_dist x y = card {i // x i ≠ y i} := rfl
+
+lemma ham_dist_comm (x y : ham β) : ham_dist x y = ham_dist y x :=
+by simp_rw [ham_dist_eq, ne_comm]
+
+lemma ham_dist_triangle (x y z : ham β) : ham_dist x z ≤ ham_dist x y + ham_dist y z :=
+begin
+  simp_rw ham_dist_eq,
+  refine le_trans (card_subtype_mono _ _ (λ _ h, _)) (card_subtype_or _ _),
+  by_contra' H, exact h (eq.trans H.1 H.2)
+end
+
+lemma ham_dist_eq_zero {x y : ham β} : ham_dist x y = 0 ↔ x = y :=
+begin
+  rw [function.funext_iff, ham_dist_eq, card_eq_zero_iff],
+  exact ⟨ λ h i, imp_of_not_imp_not _ _ (λ H, h.elim' ⟨i, H⟩) h,
+          λ h, subtype.is_empty_of_false (λ i H, H (h _))⟩
+end
+
+@[simp] lemma ham_dist_self (x : ham β) : ham_dist x x = 0 := ham_dist_eq_zero.mpr rfl
+
+lemma eq_of_ham_dist_eq_zero (x y : ham β) :
+ham_dist x y = 0 → x = y := ham_dist_eq_zero.mp
+
+lemma ham_dist_ne_zero {x y : ham β} : ham_dist x y ≠ 0 ↔ x ≠ y :=
+not_iff_not.mpr ham_dist_eq_zero
+
+lemma ham_dist_pos {x y : ham β} : 0 < ham_dist x y ↔ x ≠ y :=
+by rw [←ham_dist_ne_zero, iff_not_comm, not_lt, nat.le_zero_iff]
+
+@[simp] lemma ham_dist_eq_zero_iff_forall_eq {x y : ham β} :
+ham_dist x y = 0 ↔ ∀ i, x i = y i := by rw [ham_dist_eq_zero, function.funext_iff]
+
+lemma ham_dist_ne_zero_iff_exists_ne {x y : ham β} :
+ham_dist x y ≠ 0 ↔ ∃ i, x i ≠ y i := by rw [ham_dist_ne_zero, function.ne_iff]
+
+section has_zero
+
+variable [Π i, has_zero (β i)]
+
+/--
+The Hamming weight function to the naturals.
+-/
+
+def ham_wt (x : ham β) : ℕ := ham_dist x 0
+
+lemma ham_wt_eq (x : ham β) : ham_wt x = card {i // x i ≠ 0} := rfl
+
+lemma ham_wt_eq_ham_dist_zero (x : ham β) : ham_wt x = ham_dist x 0 := rfl
+
+lemma ham_wt_smul_le [has_zero α] [Π i, smul_with_zero α (β i)] (k : α) (x : ham β) :
+ham_wt (k • x) ≤ ham_wt x :=
+by convert ham_dist_smul_le k x _; rw smul_zero'; refl
+
+lemma ham_wt_smul [has_zero α] [Π i, smul_with_zero α (β i)] {k : α}
+(hk : ∀ i, is_smul_regular (β i) k) (x : ham β) : ham_wt x = ham_wt (k • x) :=
+by convert ham_dist_smul hk _ _; rw smul_zero'; refl
+
+lemma ham_wt_eq_zero {x : ham β} : ham_wt x = 0 ↔ x = 0 := ham_dist_eq_zero
+
+@[simp] lemma ham_wt_zero : ham_wt (0 : ham β) = 0 := ham_dist_self _
+
+lemma zero_of_ham_wt_eq_zero (x : ham β) :
+ham_wt x = 0 → x = 0 := eq_of_ham_dist_eq_zero _ _
+
+lemma ham_wt_ne_zero {x : ham β} : ham_wt x ≠ 0 ↔ x ≠ 0 := ham_dist_ne_zero
+
+lemma ham_wt_pos {x : ham β} : 0 < ham_wt x ↔ x ≠ 0 := ham_dist_pos
+
+@[simp] lemma ham_wt_zero_iff_forall_zero {x : ham β} : ham_wt x = 0 ↔ ∀ i, x i = 0 :=
+ham_dist_eq_zero_iff_forall_eq
+
+lemma ham_wt_pos_iff_exists_nz {x : ham β} : ham_wt x ≠ 0 ↔ ∃ i, x i ≠ 0 :=
+ham_dist_ne_zero_iff_exists_ne
+
+end has_zero
+
+lemma ham_dist_eq_ham_wt_sub [Π i, add_group (β i)] (x y : ham β) :
+ham_dist x y = ham_wt (x - y) :=
+by simp_rw [ham_wt_eq, ham_dist_eq, pi.sub_apply, sub_ne_zero]
```

Comment:
I think it might be better to not use the `ham` type synonym at all in this section, and reserve `ham` for use as glue between this API and the `norm`/`dist` API.

Reply:
Certainly see the logic of that.